contracts/AelinUpFrontDeal.sol

```diff
@@ -0,0 +1,299 @@
+
+        require(
+            IERC20(_dealData.underlyingDealToken).balanceOf(msg.sender) >= _underlyingDealTokenAmount,
+            "not enough balance"
+        );
+        require(_currentDealTokenTotal < _dealData.maxDealTotalSupply, "already deposited the total");
+
+        IERC20(_dealData.underlyingDealToken).transferFrom(msg.sender, address(this), _underlyingDealTokenAmount);
+
+        if (_underlyingDealTokenAmount + _currentDealTokenTotal < _dealData.maxDealTotalSupply) {
+            currentDealTokenTotal += _underlyingDealTokenAmount;
+            emit DepositDealToken(_dealData.underlyingDealToken, msg.sender, _underlyingDealTokenAmount);
+        }
+
+        if (_underlyingDealTokenAmount + _currentDealTokenTotal >= _dealData.maxDealTotalSupply) {
+            dealDepositComplete = true;
+            emit DealFullyFunded(address(this), block.timestamp, block.timestamp + _dealData.proRataRedemptionPeriod);
+        }
+    }
+
+    function purchasePoolAndAccept(IAelinPool.NftPurchaseList[] calldata _nftPurchaseList, uint256 _purchaseTokenAmount)
+        external
+    {
+        UpFrontPool memory _poolData = poolData;
+        require(block.timestamp < poolStart + _poolData.purchaseDuration, "not in purchase window");
+        require(dealDepositComplete, "deal token not yet deposited");
+        require(!poolDepositComplete, "pool completed");
+        require(IERC20(_poolData.purchaseToken).balanceOf(msg.sender) >= _purchaseTokenAmount, "not enough purchaseToken");
+
+        if (_nftPurchaseList.length > 0) {
+            _purchasePoolTokensWithNft(_poolData, _nftPurchaseList, _purchaseTokenAmount);
+        } else {
+            _purchasePoolTokens(_poolData, _purchaseTokenAmount);
+        }
+
+        totalAmountAccepted += _purchaseTokenAmount;
+        amountAccepted[msg.sender] += _purchaseTokenAmount;
+    }
+
+    function _purchasePoolTokens(UpFrontPool memory _poolData, uint256 _purchaseTokenAmount) internal {
+        if (hasAllowList) {
+            require(_purchaseTokenAmount <= allowList[msg.sender], "more than allocation");
+            allowList[msg.sender] -= _purchaseTokenAmount;
+        }
+
+        uint256 currentBalance = IERC20(_poolData.purchaseToken).balanceOf(address(this));
+        IERC20(_poolData.purchaseToken).transferFrom(msg.sender, address(this), _purchaseTokenAmount);
+        uint256 balanceAfterTransfer = IERC20(_poolData.purchaseToken).balanceOf(address(this));
+        uint256 purchaseTokenAmount = balanceAfterTransfer - currentBalance;
+
+        if (_poolData.purchaseTokenCap > 0) {
+            uint256 totalPoolAfter = totalSupply() + purchaseTokenAmount;
+            require(totalPoolAfter <= _poolData.purchaseTokenCap, "cap has been exceeded");
+            if (totalPoolAfter == _poolData.purchaseTokenCap) {
+                poolDepositComplete = true;
+            }
+        }
+
+        _mint(msg.sender, purchaseTokenAmount);
```

Comment:
Here I was thinking that we make the pool tokens quantity 1:1:1 with the deal tokens and underlying deal tokens. So we should have a uint256 in the interface for purchaseTokensPerDealTokens that sets the price of the deal tokens, instead of minting the purchase token amount from the input, it mints _purchaseTokenAmount / purchaseTokensPerDealTokens. Then this purchaseTokensPerDealTokens is what we can vary when for the dutch auction aelip.

Reply:
Since this is all done in this one contract, what do you think about holding this value for pool tokens in a mapping instead of minting an erc20, then convert that to the deal token erc20 when they claim their deal tokens after the minimum raise threshold is passed and the underlying deal tokens have been deposited? This gives a good point to apply the aelin fee escrow for the treasury and the sponsor fee.

Reply:
let's discuss these 2 comments in standies tomorrow